### PR TITLE
Scoring: Implement a registry of Score classes

### PR DESCRIPTION
**Pull request by janezd**
##### Issue

We want to let add-ons add new classes for scoring models. The Test & Score widget should recognize and use them.

##### Description of changes

- PR implements a registry of subclasses of `Score` through meta class `ScoreMetaType`. It is derived from `WrapperMeta` and implements a registry. Multiple inheritance (from `Registry`) can't be used since a class can't have two `__new__` methods.

- Subclasses of `Score` can provide a better name and a tooltip for the column. This is undesirable (the code on this level shouldn't be aware of GUI) but rather inevitable if we want to let add-ons add stuff to widgets.

- Subclasses of `Score` can be declared abstract (see `ClassificationScore`, `RegressionScore`)

- The Test & Score widget takes all non-abstract scores from the registry. Built-in scores are ordered as they were before this PR.

Since the PR didn't pass the coverage test because of reformatting `graph_ranks`, I now call this function from tests. It's difficult to test what it draws, so the tests just ensures it doesn't straight out crash.

##### Includes
- [X] Code changes
- [x] Tests

<strike>I don't know how and what to test here. Testing `owtestlearners._get_score` on patched `scoring.Score.registry` seems ridiculous.</strike>

### CODE REVIEW

File: Orange/evaluation/scoring.py [hunk 1]
```diff
@@ -287,46 +286,35 @@ def compute_score(self, results, eps=1e-15, normalize=True,
...
-
-
 # pylint: disable=invalid-name
-class R2(Score):
+class R2(RegressionScore):
     __wraps__ = skl_metrics.r2_score
-    class_types = (ContinuousVariable, )
     long_name = "Coefficient of determination"
 
     def compute_score(self, results):
```
Comment (janezd) [inline Orange/evaluation/scoring.py]: This method is removed in the next comment. (Adding by chunks didn't work properly. :))

**Comment by astaric** [comment]
You can test the MetaClass magic:
 - passing name as a keyword argument
 - name on inherited classes
And widget
 - is "DummyScore" defined in a test visible in Test & Score columns

Also, please rebase, as I have merged the other Test & Score PR.

File: Orange/evaluation/scoring.py [hunk 2]
```diff
@@ -11,26 +11,67 @@
...
+
+    def __init__(cls, *args, **_):
+        WrapperMeta.__init__(cls, *args)
+
+
+# Derived classes may include a keyword argument name with a short readable
+# name of the score, and an attribute long_name with a few word description.
+#
+# class TGOGScore(Score, name="Thingamagog"):
+#    long_name = "Magic Thingamagog score"
```
Comment (astaric) [inline Orange/evaluation/scoring.py]: TypeError: `__new__()` got multiple values for argument 'name'

File: Orange/widgets/evaluate/owtestlearners.py [hunk 3]
```diff
@@ -538,19 +548,18 @@ def puts_results(self, learners, results, class_var):
...
-            else:
-                headers.extend(regression_stats.headers)
-
-        # remove possible extra columns from the model.
-        for i in reversed(range(len(headers),
-                                self.result_model.columnCount())):
-            self.result_model.takeColumn(i)
-
-        self.result_model.setHorizontalHeaderLabels(headers)
+            func = classification_stats if self.data.domain.has_discrete_class \
```
Comment (astaric) [inline Orange/widgets/evaluate/owtestlearners.py]: I would prefer

```
if self.data.domain.has_discrete_class:
    scores = classification_stats.scores
else:
    scores = regression_stats.scores

File: Orange/widgets/evaluate/owtestlearners.py [hunk 4]
```diff
@@ -538,19 +548,18 @@ def puts_results(self, learners, results, class_var):
...
+                else regression_stats
+            scores = func.scores
+        else:
+            scores = []
+        model = self.result_model
+        model.setColumnCount(1 + len(scores))
+        for col_1, score in enumerate(scores):
```
Comment (astaric) [inline Orange/widgets/evaluate/owtestlearners.py]: what does col_1 stand for? Would name index/idx suffice?

**Comment by kernc** [comment]
> I don't know how and what to test here.

You can use the registry to test all scorers in a single place with something meaningful and/or edge-cases.

File: Orange/evaluation/scoring.py (hunk 2 above)
Comment (janezd) [inline Orange/evaluation/scoring.py]: The comment referred to an older, broken code. `name` is now just a class attribute. I removed the comment.

File: Orange/widgets/evaluate/owtestlearners.py (hunk 3 above)
Comment (janezd) [inline Orange/widgets/evaluate/owtestlearners.py]: Fixed: this code is gone altogether.

File: Orange/widgets/evaluate/owtestlearners.py (hunk 4 above)
Comment (janezd) [inline Orange/widgets/evaluate/owtestlearners.py]: It stood for `col - 1`. Apparently only I use this convention, so I removed `_1`.

**Comment by janezd** [comment]
> You can test the MetaClass magic:
> - passing name as a keyword argument

It wasn't intended to work like that. The comment about it remained from an older, broken version.

> name on inherited classes

Check.

> And widget: is "DummyScore" defined in a test visible in Test & Score columns

It was once again a good idea to force people (=me) to test their code. :)

A list of scores was constructed when module was loaded, so it would appear only if the add-on was loaded before the widget. And it could never work in tests. In the process of fixing this, I removed functions `classification_stats` and `regression_stats` that were never called, but were used for storing attribute `scores`. The code affected by their removal is now simpler than before.

**Comment by janezd** [comment]
@kernc 

> You can use the registry to test all scorers in a single place with something meaningful and/or edge-cases.

This is a good idea. Also, the widget could use some additional tests. But not in this PR.

**Comment by astaric** [comment]
@janezd,this PR requires a non-trivial rebase onto the modified test&score widget. Please mention me when done so I can merge.
